Miscellaneous Fixes

hbhanawat opened the pull request:
## Changes proposed in this pull request

There are two changes:
1. Added a HeapDumper class that is used to dump the heap. The dumpHeap function gets called when the ForcedDisconnectException is generated. But the heap is generated only when the property gemfire.dumpheap is set as true. This is being introduced as debugging feature for our test environments.
2. Made fixes to the logging as per the comments on SNAP-1420.
3. Fixed a bug which was causing issues if the snappy_hive_metastore was the last schema during ddl replay. The level was set as WARN for always.


## Patch testing
Manual testing of heap dumper and logging related changes. 
Precheckin.

Commit:
There are two changes:
1. Added a HeapDumper class that is used to dump the heap. The dumpHeap function gets called when the ForcedDisconnectException is generated. But the heap is generated only when the property gemfire.dumpheap is set as true. This is being introduced as debugging feature for our test environments.
2. Made fixes to the logging as per the comments on SNAP-1420.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/HeapDumper.java:
```diff
@@ -0,0 +1,76 @@
+    private static volatile HotSpotDiagnosticMXBean hotspotMBean;
+
+    static void dumpHeap(boolean live) {
+        boolean dump = Boolean.getBoolean("gemfire.dumpheap");
```
Can you change to SystemProperties.getServiceInstance().getBoolean? That allows to switch the prefix as per the product being used (and we can add "snappydata.store" there too).

hbhanawat replied:
I made the ServerInstance change. But, I don't want to add snapydata.store prefix in the callbacks as it may impact existing snappydata hydra tests. Is it ok to leave it as it is for now and add this comment to our property spec?

hbhanawat replied:
Updated the property spec. See point 8. 

https://docs.google.com/document/d/1AhcvGHuB0VodV2TNh27KE_9kvlzxaj42gY8rQ_Dr3IY/edit

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/HeapDumper.java:
```diff
@@ -0,0 +1,76 @@
+        if (!dump) {
+            return;
+        }
+        String fileName = "heap" + System.currentTimeMillis() + ".bin";
```
Can you add protecting this so that only one heap dump happens in some time duration? Perhaps ForceDisconnect may not be invoked from multiple threads, but as a general API it is better to have the contract that this will not dump multiple times close together. For stack dumps there was a time limit e.g. make this method synchronized, record this System.currentTimeMillis() in a static long, then check against it that previous one should not be too close like < 3 seconds else skip the dump.

hbhanawat replied:
Done.

sumwale reviewed (changes requested):
Overall looks good. See couple of comments.

Commit:
Incorporated review comments
Fixed a bug which was causing issues if the snappy_hive_metastore was the last schema during ddl replay. The level was set as WARN for always.

Commit:
Merge branch 'snappy/master' into MiscFixes

Commit:
Merge branch 'snappy/master' into MiscFixes

Conflicts:
	gemfire-shared/src/main/resources/store-log4j.properties
